src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+    std::unique_ptr<JITEventListener> JuliaListener;
+
+public:
+    DebugObjectRegistrar() : JuliaListener(CreateJuliaJITEventListener()) {}
+
+    template <typename ObjSetT, typename LoadResult>
+    void operator()(ObjectLinkingLayerBase::ObjSetHandleT, const ObjSetT &Objects,
+                  const LoadResult &LOS) {
+        auto oit = Objects.begin();
+        auto lit = LOS.begin();
+        while (oit != Objects.end()) {
+            auto &Object = *oit;
+            auto &LO = *lit;
+
+            OwningBinary<ObjectFile> SavedObject = LO->getObjectForDebug(*Object);
+
+            // If the debug object is unavailable, save (a copy of) the original object
+            // for our backtraces
+            if (!SavedObject.getBinary()) {
+                // This is unfortunate, but there doesn't seem to be a way to take
+                // ownership of the original buffer
+                auto NewBuffer = MemoryBuffer::getMemBufferCopy(Object->getData(), Object->getFileName());
+                auto NewObj = ObjectFile::createObjectFile(NewBuffer->getMemBufferRef());
+                SavedObject = OwningBinary<ObjectFile>(std::move(*NewObj),std::move(NewBuffer));
+            }
+            else
+                NotifyGDB(SavedObject);
+
+            SavedObjects.push_back(std::move(SavedObject));
+            ORCNotifyObjectEmitted(JuliaListener.get(),*Object,*SavedObjects.back().getBinary(),*LO);
+
+            ++oit;
+            ++lit;
+        }
+    }
+};
+
+}
+
+#if defined(_OS_DARWIN_) && defined(LLVM37) && defined(LLVM_SHLIB)
+#define CUSTOM_MEMORY_MANAGER 1
+extern RTDyldMemoryManager* createRTDyldMemoryManagerOSX();
+#endif
+
+class JuliaOJIT {
+public:
+    typedef orc::ObjectLinkingLayer<DebugObjectRegistrar> ObjLayerT;
+    typedef orc::IRCompileLayer<ObjLayerT> CompileLayerT;
+    typedef CompileLayerT::ModuleSetHandleT ModuleHandleT;
+    typedef StringMap<void*> GlobalSymbolTableT;
+    typedef object::OwningBinary<object::ObjectFile> OwningObj;
+
+    JuliaOJIT(TargetMachine &TM)
+      : TM(TM),
+        DL(TM.createDataLayout()),
+        ObjStream(ObjBufferSV),
+        MemMgr(
+#ifdef CUSTOM_MEMORY_MANAGER
+            createRTDyldMemoryManagerOSX()
+#else
+            new SectionMemoryManager
+#endif
+            ) {
+#ifdef JL_DEBUG_BUILD
+            PM.add(createVerifierPass());
+#endif
+            // In imaging mode, we run the pass manager on creation
+            // to make sure it ends up optimized in the shadow module
+            if (!imaging_mode) {
+                addOptimizationPasses(&PM);
+#ifdef JL_DEBUG_BUILD
+                PM.add(createVerifierPass());
+#endif
+            }
+            if (TM.addPassesToEmitMC(PM, Ctx, ObjStream))
+                llvm_unreachable("Target does not support MC emission.");
+
+            CompileLayer = std::unique_ptr<CompileLayerT>{new CompileLayerT(ObjectLayer,
+                [&](Module &M) {
+                    PM.run(M);
```

Comment:
The plan is to not use ORC on 3.7.1 (without patches) and continue to use MCJIT there.